ffi/derive/src/bindgen.rs

```diff
@@ -272,17 +273,34 @@ fn gen_ptr_null_check_stmt(fn_arg_descriptor: &FnArgDescriptor) -> Option<syn::S
     None
 }
 
+fn gen_dangling_ptr_assignments(fn_descriptor: &FnDescriptor) -> Vec<syn::Stmt> {
+    let mut stmts = vec![];
+
+    for arg in &fn_descriptor.input_args {
+        if arg.is_slice_ref() {
+            stmts.push(gen_dangling_ptr_assignment(arg));
+        }
+    }
+    if let Some(output_arg) = ffi_output_arg(fn_descriptor) {
+        if output_arg.is_slice_ref_mut() {
+            stmts.push(gen_dangling_ptr_assignment(output_arg));
+        }
+    }
+
+    stmts
+}
+
+// NOTE: `slice::from_raw_parts` takes a non-null aligned pointer
 fn gen_dangling_ptr_assignment(fn_arg_descriptor: &FnArgDescriptor) -> syn::Stmt {
     let (arg_name, slice_len_arg_name) = (
         &fn_arg_descriptor.ffi_name,
         gen_slice_len_arg_name(fn_arg_descriptor),
     );
 
     parse_quote! {
-        if #slice_len_arg_name == 0_usize {
-            // NOTE: `slice::from_raw_parts` takes a non-null aligned pointer
-            #arg_name = core::ptr::NonNull::dangling().as_ptr();
-        }
+        let #arg_name = if #slice_len_arg_name == 0_usize {
+            core::ptr::NonNull::dangling().as_ptr()
```

Comment:
It's up to the macro composer to properly initialize this pointer, right?

Reply:
I would say yes. Consider the following example:

```rs
fn call_slicer() {
	let arr: [u8; 0] = [];

	// arr.as_ptr doesn't return null pointer. Therefore, it can be used with `core::slice::from_raw_parts`
	slicer(arr.as_ptr(), arr.len());
	
	// if null pointer is given then it cannot be used to construct zero-sized slice with `core::slice::from_raw_parts`
	// since this API is quite natural for a user who wishes to skip the use of the slice `slice_ptr` must be assigned non-null dangling value
	slicer(core::ptr::null(), 0);
}

extern fn slicer(slice_ptr: *const u8, slice_len: usize);
```

Reply:
Okay, 0-length slices are interpreted as dangling pointers and then always restored by `slice::from_raw_parts`, both in arguments and in return values

Reply:
it would be a logical error to call `slice::from_raw_parts` with null pointer

Reply:
So we assign dangling pointers to 0-length slices so as not to be considered null